### ISSUE TITLE
Error while downloading tar/zip file from s3 fileserver

### ISSUE THREAD

**Issue by xoraes** [saltstack/salt#23567]
Salt failed with this stacktrace when downloading a tar file from s3. 

I ran the following:
```yaml
unzip_file:
  file.managed:
    - mode: 755
    - source: salt://s3/de/dev/de.zip
    - name: /tmp/de.zip
```
```python
2015-05-10 07:42:05,599 [salt.daemons.masterapi][ERROR   ] Exception need more than 1 value to unpack occurred in file server update
"master" 2790L, 360113C                                                                                                                                                                                                   1,1           Top
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/__init__.py", line 512, in file_hash
    fnd = self.find_file(load['path'], load['saltenv'])
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/__init__.py", line 466, in find_file
    fnd = self.servers[fstr](path, saltenv, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/s3fs.py", line 152, in find_file
    _get_file_from_s3(metadata, saltenv, fnd['bucket'], path, cached_file_path)
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/s3fs.py", line 585, in _get_file_from_s3
    name, value = header.split(':', 1)
ValueError: need more than 1 value to unpack
2015-05-12 05:20:07,923 [salt.daemons.masterapi][ERROR   ] Exception need more than 1 value to unpack occurred in file server update
2015-05-12 05:21:13,496 [salt.daemons.masterapi][ERROR   ] Exception need more than 1 value to unpack occurred in file server update
2015-05-12 05:22:19,094 [salt.daemons.masterapi][ERROR   ] Exception need more than 1 value to unpack occurred in file server update
2015-05-12 05:22:45,615 [salt.master      ][ERROR   ] Error in function _file_hash:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1348, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/__init__.py", line 512, in file_hash
    fnd = self.find_file(load['path'], load['saltenv'])
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/__init__.py", line 466, in find_file
    fnd = self.servers[fstr](path, saltenv, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/s3fs.py", line 152, in find_file
    _get_file_from_s3(metadata, saltenv, fnd['bucket'], path, cached_file_path)
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/s3fs.py", line 585, in _get_file_from_s3
    name, value = header.split(':', 1)
ValueError: need more than 1 value to unpack
```

I am using the following version of salt:
```yaml
                  Salt: 2014.7.5
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
 Debian source package: 2014.7.5+ds-1ubuntu1
```

Any suggestions?

**Comment by jfindlay** [comment]
@xoraes, this looks to be a bug to me.  Have you tried 2015.5.0?

**Comment by xoraes** [comment]
No, we haven't tried 2015.5.0.

**Comment by dennisoconnor** [comment]
I can confirm the same error on 2015.5.0.

**Comment by sebest** [comment]
Any update on fixing this regression?

It appeared sometimes during 2014.7.x with x>1

**Comment by rdodev** [comment]
FWIW it seems to work correctly, at least for me, in version: 2015.5.2 (Lithium)

**Comment by sebest** [comment]
Still not available as a debian package?!

**Comment by rallytime** [comment]
@xoraes Thanks for the confirmation!

@sebest The 2015.5.2 release has not yet been announced officially. We only pushed the tag today to the salt repo and have just recently notified our community packagers. Packages should be available around the time we announce the release next week.

**Comment by sebest** [comment]
@rallytime it would be really convenient to have more frequent bugfix release especially when we are talking about regression, for this specific regression there is a 1 month delay, so for the servers that we upgraded from 2014.7.x it is really troublesome.

About the packaging as far as i know there was no 2015.5.1 debian package.

**Comment by basepi** [comment]
@sebest We've been targeting 2 week bugfix releases for the last couple of releases. However, that may change because our community packagers may not be able to keep up with that frequency. In any case, bugfix releases will definitely happen more frequently than they used to.

**Comment by sebest** [comment]
@basepi that's really good to know!

I thought saltstack planned to manage packaging internally for the major linux distributions?

**Comment by basepi** [comment]
Yep, we're still working on executing those plans. But we're also working on shipping an awesome enterprise UI, and keeping up with an awesome community. So much work, so little time! =)

**Comment by sebest** [comment]
@basepi I'm glad that salt will soon have an awesome enterprise UI, but focusing on stability and fixing regression and shipping bugfix release is critical for enterprises (more than having a UI).
We are currently affected by 2 regressions: this one and https://github.com/saltstack/salt/issues/23822

The bugs have been opened for a month.

On a side note i can see that salt v2015.5.1 was released 29 days ago but no Ubuntu/Debian package is available, i think that Ubuntu/Debian are major Linux distributions, and if 2015.5.1 is only a bugfix release of 2015.5.0, it should not take a month to release updated packages.

We are using salt-cloud to deploy our servers and i have expecting to be able to use salt-bootstrap to deploy from github so we would be able to deploy a version of with our fixes from a github fork, but for some reason salt-cloud / salt-boostrap insists on deploying the outdated packages, i'm using the script arguments -P git v2015.5.1 , but it still deploy the debian packages.

**Comment by pcn** [comment]
As an added data point, I'm testing out s3fs and I'm getting this with the ubuntu ppa for 2015.5.2.

On the saltmaster:

```
ubuntu@peter.norton-saltmaster-vagrant:/srv/salt/states/cassandra$ sudo salt saltmaster test.versions_report
saltmaster:
                      Salt: 2015.5.2
                    Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
                    Jinja2: 2.7.2
                  M2Crypto: 0.21.1
            msgpack-python: 0.3.0
              msgpack-pure: Not Installed
                  pycrypto: 2.6.1
                   libnacl: Not Installed
                    PyYAML: 3.10
                     ioflo: Not Installed
                     PyZMQ: 14.0.1
                      RAET: Not Installed
                       ZMQ: 4.0.4
                      Mako: 0.9.1
     Debian source package: 2015.5.2+ds-1trusty1
```

On the minion:

```
ubuntu@peter.norton-saltmaster-vagrant:/srv/salt/states/cassandra$ sudo salt basenode test.versions_report
basenode:
                      Salt: 2015.5.2
                    Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
                    Jinja2: 2.7.2
                  M2Crypto: 0.21.1
            msgpack-python: 0.3.0
              msgpack-pure: Not Installed
                  pycrypto: 2.6.1
                   libnacl: Not Installed
                    PyYAML: 3.10
                     ioflo: Not Installed
                     PyZMQ: 14.0.1
                      RAET: Not Installed
                       ZMQ: 4.0.4
                      Mako: 0.9.1
     Debian source package: 2015.5.2+ds-1trusty1
```

**Comment by pcn** [comment]
I can't get the code from devel or from branches post-2015.05.xy to work. https://github.com/saltstack/salt/commit/8999148da0a96145f289d1049802e49c5d862339 seems to be in the 2015.5.2 ubuntu package.  https://github.com/saltstack/salt/commit/80c7a697ae10b63e807d1b44755ae1338b306cd1 does not appear to be there.  I've got a patch I'll link to here that works for me that change conventions of s3.query based on the conventions that the latter patch seems to bring in, like lowercase headers (per the amazon auth header requirements IIRC).

**Comment by rallytime** [comment]
@pcn @xoraes I have back-ported the fix from @pcn and the other commits that were referenced in #25576 to 2015.5. The relevant back-port PR is #25584. Can you guys try that at the HEAD of the 2015.5 branch when #25584 gets merged and confirm that this is fixed? Thanks in advance! :)

**Comment by sebest** [comment]
@rallytime i tried getting the files from your pull request and use them with 2015.5.3 and ran into some issue.
It seems that we need part of this commit: https://github.com/rallytime/salt/commit/c63c2356be0bbf7e52d9e642cb3ed055323bd7bf
the function sig4 has a new parameter 'data' and aws_api_version is optional with a default value.

using the changes in the mention commit for salt/utils/aws.py solved my issue.

thanx

**Comment by rallytime** [comment]
@sebest Good info! The pull request containing that commit, #25420, was made against 2015.5 and has been merged forward to 2015.8 and develop. (And if it hasn't made it all the way there yet, it will soon.) So this bug should be fixed in the next release of Salt. Given that information, is this good to close?

**Comment by sebest** [comment]
@rallytime in my opinion, every regressions introduced in a major release should be fixed in a point release  of that major release as production environment can't upgrade to next major as easily:
- release date of major release often slipped to a couple weeks (months) later
- packaging (ubuntu/debian) is lagging (often 1 month after the official release)
- new regressions are introduced (like this one)

I've been using salt since 0.16 and stability is still the main concern that we have with new release.
We migrate from 2014.7 to 2015.5 and 2 regressions are still open as of 2015.5.3.

So you can close it but it is important for users to know that 2015.5.x will never see a fix for this issue and must upgrade to 2015.8.x

I really like the product but improving stability and fixing regressions and releasing updated packages in a timely manner should be high priority.

**Comment by rallytime** [comment]
You bring up very fair and valid points and we are working very hard to make Salt more stable. In terms of this bug being fixed, 2014.5.x *will* see another release that includes this fix. We will be releasing 2015.5.4, though I am not sure on the date. I was being vague on my last comment about `should be fixed in the next release of Salt`. What I meant by that was both 2015.5.4 and 2015.8.0 will see these fixes, but I am not certain which release will be packaged up first. 

The problems you've mentioned above about our packaging and release cycle are at the top of our list. Some of the things we are working very hard on are getting package building to be an automated process (we currently rely on community packagers for many of our packages), increasing the frequency of our release cadence, and improving the quality of our releases so that fewer regressions are produced. 

We are actively working towards all of these issues and they are certainly high priority and we definitely appreciate your comments.
